File: app/src/main/java/com/xephorium/crystalnote/ui/widget/NotesWidgetProvider.java
package com.xephorium.crystalnote.ui.widget;

import android.app.PendingIntent;
import android.appwidget.AppWidgetManager;
import android.appwidget.AppWidgetProvider;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.widget.RemoteViews;
import android.widget.Toast;

import com.xephorium.crystalnote.R;
import com.xephorium.crystalnote.data.NoteRepository;
import com.xephorium.crystalnote.data.SharedPreferencesManager;
import com.xephorium.crystalnote.data.model.Note;
import com.xephorium.crystalnote.data.util.NoteUtils;
import com.xephorium.crystalnote.ui.IntentLibrary;
import com.xephorium.crystalnote.ui.selection.SelectionActivity;

import java.util.List;

/*
  NotesWidgetProvider.java                     04.15.2017
  Christopher Cruzen

    Defines appearance and view behavior of notes widget.
*/

public class NotesWidgetProvider extends AppWidgetProvider {

    private static final String BUTTON_CLICK_INTENT = "com.xephorium.crystalnote.widget.click.BUTTON";
    private static final String TEXT_CLICK_INTENT = "com.xephorium.crystalnote.widget.click.TEXT";
    private static final String TITLE_CLICK_INTENT = "com.xephorium.crystalnote.widget.click.TITLE";

	@Override
	public void onUpdate(Context context, AppWidgetManager appWidgetManager, int[] appWidgetIds) {
		int[] widgetInstances = getWidgetInstances(context, appWidgetManager);

		for (int widgetInstance : widgetInstances) {

            RemoteViews remoteViews = new RemoteViews(context.getPackageName(), R.layout.note_widget_layout);
            NoteRepository noteRepository = new NoteRepository(context);
            List<Note> availableNotes = noteRepository.getNotes();
            Note displayNote = NoteUtils.getNoteFromList(availableNotes, SharedPreferencesManager.getDisplayNoteName(context));

            if (displayNote != null) {
                remoteViews.setTextViewText(R.id.note_widget_title, displayNote.getName());
                remoteViews.setTextViewText(R.id.note_widget_text, noteRepository.getNoteContents(noteRepository
                        .getNoteFile(displayNote)));

            } else if (availableNotes != null && availableNotes.size() > 0) {
                remoteViews.setTextViewText(R.id.note_widget_title, availableNotes.get(0).getName());
                remoteViews.setTextViewText(R.id.note_widget_text, noteRepository.getNoteContents(noteRepository
                        .getNoteFile(availableNotes.get(0))));
            } else {
                remoteViews.setTextViewText(R.id.note_widget_title, "");
                remoteViews.setTextViewText(R.id.note_widget_text, "");
            }

            remoteViews.setOnClickPendingIntent(R.id.note_widget_button, getOnClickPendingIntent(context, BUTTON_CLICK_INTENT));
            remoteViews.setOnClickPendingIntent(R.id.note_widget_text, getOnClickPendingIntent(context, TEXT_CLICK_INTENT));
            remoteViews.setOnClickPendingIntent(R.id.note_widget_title, getOnClickPendingIntent(context, TITLE_CLICK_INTENT));

			appWidgetManager.updateAppWidget(widgetInstance, remoteViews);
		}
	}

    @Override
    public void onReceive(Context context, Intent intent) {
        super.onReceive(context, intent);

        switch (intent.getAction()) {
            case BUTTON_CLICK_INTENT:
                Intent buttonIntent = new Intent(context, SelectionActivity.class);
                buttonIntent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TASK);
                buttonIntent.setAction(IntentLibrary.CHOOSE_NOTE_INTENT);
                context.startActivity(buttonIntent);
                break;

            case TEXT_CLICK_INTENT:
                Toast.makeText(context, "Text Clicked", Toast.LENGTH_SHORT).show();
                break;

            case TITLE_CLICK_INTENT:
                Toast.makeText(context, "Title Clicked", Toast.LENGTH_SHORT).show();
                break;

            case IntentLibrary.UPDATE_NOTE_INTENT:
                updateWidgets(context);
                break;
        }
    }

    private PendingIntent getOnClickPendingIntent(Context context, String intentAction) {
        Intent intent = new Intent(context, this.getClass());
        intent.setAction(intentAction);
        return PendingIntent.getBroadcast(context, 0, intent, 0);
    }

    private void updateWidgets(Context context) {
        AppWidgetManager appWidgetManager = AppWidgetManager.getInstance(context.getApplicationContext());
        int[] widgetInstances = getWidgetInstances(context, appWidgetManager);

        if (widgetInstances != null && widgetInstances.length > 0)
            onUpdate(context, appWidgetManager, widgetInstances);
    }

    private int[] getWidgetInstances(Context context, AppWidgetManager appWidgetManager) {
        ComponentName widgetName = new ComponentName(context, this.getClass());
        return appWidgetManager.getAppWidgetIds(widgetName);
    }
}

File: app/src/main/java/com/xephorium/crystalnote/ui/widget/NotesWidgetProvider.kt
package com.xephorium.crystalnote.ui.widget

import android.app.Application
import android.app.PendingIntent
import android.appwidget.AppWidgetManager
import android.appwidget.AppWidgetProvider
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.util.TypedValue.COMPLEX_UNIT_SP
import android.view.View
import android.widget.RemoteViews

import com.xephorium.crystalnote.R
import com.xephorium.crystalnote.data.repository.NoteRoomRepository
import com.xephorium.crystalnote.data.repository.SharedPreferencesRepository
import com.xephorium.crystalnote.ui.select.SelectActivity
import java.util.concurrent.Callable
import java.util.concurrent.Executors
import com.xephorium.crystalnote.data.model.WidgetState
import com.xephorium.crystalnote.data.utility.ColorUtility
import com.xephorium.crystalnote.data.utility.CrystalNoteLogger
import com.xephorium.crystalnote.ui.update.UpdateNoteActivity
import com.xephorium.crystalnote.ui.update.UpdateNoteActivity.Companion.KEY_LAUNCH_FROM_WIDGET
import com.xephorium.crystalnote.ui.update.UpdateNoteActivity.Companion.KEY_NOTE_ID


/*
  NotesWidgetProvider                                      05.11.2019
  Christopher Cruzen

    Defines appearance and view behavior for all CrystalNote widgets.
*/

class NotesWidgetProvider : AppWidgetProvider() {


    /*--- Lifecycle Methods ---*/

    // Handle Widget Update
    override fun onUpdate(
        context: Context,
        widgetManager: AppWidgetManager,
        widgetIdsPendingUpdate: IntArray
    ) {

        // Create Update Variables
        val sharedPreferencesRepository = SharedPreferencesRepository(context)

        widgetIdsPendingUpdate.let { widgetIds ->

            // For Each Widget
            for (widgetId in widgetIds) {

                // Get Widget Variables
                val widgetView = RemoteViews(context.packageName, R.layout.note_widget_layout)
                val widgetState = sharedPreferencesRepository.getWidgetState(widgetId)
                val widgetNoteId = sharedPreferencesRepository.getNoteIdForWidget(widgetId)
                val widgetNote = widgetNoteId?.let { id ->
                        val callable = Callable {
                            NoteRoomRepository(context).getNoteSynchronously(id)
                        }
                        val future = Executors.newSingleThreadExecutor().submit(callable)
                        future.get()
                    }

                // Set Widget Style
                if (widgetState == null) {
                    styleWidget(widgetView, WidgetState(0, 0))
                } else {
                    styleWidget(widgetView, widgetState)
                }

                // Populate Widget Fields
                if (widgetNote != null) {

                    if (widgetNote.password.isEmpty()) {

                        // Populate Normal State
                        widgetView.setViewVisibility(R.id.textWidgetTitle, View.VISIBLE)
                        widgetView.setViewVisibility(R.id.textWidgetContent, View.VISIBLE)
                        widgetView.setViewVisibility(R.id.textWidgetEmpty, View.GONE)
                        widgetView.setViewVisibility(R.id.textWidgetLocked, View.GONE)
                        widgetView.setTextViewText(R.id.textWidgetTitle, widgetNote.name)
                        widgetView.setTextViewText(R.id.textWidgetContent, widgetNote.contents)


                    } else {

                        // Populate Locked State
                        widgetView.setViewVisibility(R.id.textWidgetTitle, View.GONE)
                        widgetView.setViewVisibility(R.id.textWidgetContent, View.GONE)
                        widgetView.setViewVisibility(R.id.textWidgetEmpty, View.GONE)
                        widgetView.setViewVisibility(R.id.textWidgetLocked, View.VISIBLE)
                    }

                } else {

                    // Populate Empty State
                    widgetView.setViewVisibility(R.id.textWidgetTitle, View.GONE)
                    widgetView.setViewVisibility(R.id.textWidgetContent, View.GONE)
                    widgetView.setViewVisibility(R.id.textWidgetEmpty, View.VISIBLE)
                    widgetView.setViewVisibility(R.id.textWidgetLocked, View.GONE)
                }

                // Set Listeners
                widgetView.setOnClickPendingIntent(
                        R.id.textWidgetTitle,
                        getOnClickPendingIntent(context, TITLE_CLICK_INTENT, widgetId)
                )
                widgetView.setOnClickPendingIntent(
                        R.id.textWidgetContent,
                        getOnClickPendingIntent(context, TEXT_CLICK_INTENT, widgetId)
                )
                widgetView.setOnClickPendingIntent(
                        R.id.textWidgetEmpty,
                        getOnClickPendingIntent(context, EMPTY_CLICK_INTENT, widgetId)
                )

                widgetManager.updateAppWidget(widgetId, widgetView)
            }
        }
    }

    // Handle Intents Generated by Widget Interaction
    override fun onReceive(context: Context, intent: Intent) {
        super.onReceive(context, intent)

        // Create Receive Variables
        val widgetId = intent.getIntExtra(KEY_WIDGET_ID, 0)

        when (intent.action) {

            TITLE_CLICK_INTENT, TEXT_CLICK_INTENT -> {

                // Launch UpdateActivity for Note
                SharedPreferencesRepository(context).getNoteIdForWidget(widgetId)?.let { id ->
                    val updateIntent = Intent(context, UpdateNoteActivity::class.java)
                    updateIntent.putExtra(KEY_NOTE_ID, id)
                    updateIntent.putExtra(KEY_LAUNCH_FROM_WIDGET, true)
                    updateIntent.addFlags(
                            Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK)
                    context.startActivity(updateIntent)
                }
            }

            EMPTY_CLICK_INTENT -> {

                // Choose New Note
                val buttonIntent = Intent(context, SelectActivity::class.java)
                buttonIntent.flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK
                buttonIntent.putExtra(KEY_WIDGET_ID, widgetId)
                context.startActivity(buttonIntent)
            }
        }
    }

    // Handle Widget Delete
    override fun onDeleted(context: Context, widgetIds: IntArray) {
        super.onDeleted(context, widgetIds)

        // Remove Widget ID from Shared Preferences
        val sharedPreferencesRepository = SharedPreferencesRepository(context)
        for (widgetId in widgetIds) {
            sharedPreferencesRepository.removeWidgetState(widgetId)
        }
    }

    // Handle Final Widget Deletion
    override fun onDisabled(context: Context) {
        super.onDisabled(context)
        SharedPreferencesRepository(context).removeAllWidgetStates()
    }

    // Handle Device Reboot
    override fun onRestored(context: Context, oldWidgetIds: IntArray, newWidgetIds: IntArray) {
        super.onRestored(context, oldWidgetIds, newWidgetIds)

        // Create Update Variables
        val sharedPreferencesRepository = SharedPreferencesRepository(context)
        val widgetStateList = sharedPreferencesRepository.getWidgetStateList()

        // Update Widget ID's
        for (index in oldWidgetIds.indices) {
            widgetStateList.updateWidgetId(oldWidgetIds[index], newWidgetIds[index])
        }
        sharedPreferencesRepository.setWidgetStateList(widgetStateList)
    }


    /*--- Private Methods ---*/

    private fun styleWidget(widgetView: RemoteViews, state: WidgetState) {

        // Background Color
        widgetView.setInt(
            R.id.imageWidgetBackground,
            "setColorFilter",
            state.backgroundColor
        )

        // Background Transparency
        widgetView.setInt(
            R.id.imageWidgetBackground,
            "setImageAlpha",
            ((1.0 - state.backgroundAlpha.value) * 255.0).toInt()
        )

        // Text Size
        widgetView.setTextViewTextSize(
            R.id.textWidgetTitle,
            COMPLEX_UNIT_SP,
            (state.textSize.size + 1).toFloat()
        )
        widgetView.setTextViewTextSize(
            R.id.textWidgetContent,
            COMPLEX_UNIT_SP,
            (state.textSize.size).toFloat()
        )
        widgetView.setTextViewTextSize(
            R.id.textWidgetEmpty,
            COMPLEX_UNIT_SP,
            (state.textSize.size).toFloat()
        )
        widgetView.setTextViewTextSize(
            R.id.textWidgetLocked,
            COMPLEX_UNIT_SP,
            (state.textSize.size).toFloat()
        )

        // Text Color
        widgetView.setTextColor(R.id.textWidgetTitle, state.titleColor)
        widgetView.setTextColor(
            R.id.textWidgetContent,
            ColorUtility.applyTransparency(state.contentColor, state.contentAlpha)
        )
        widgetView.setTextColor(
            R.id.textWidgetEmpty,
            ColorUtility.applyTransparency(state.contentColor, state.contentAlpha)
        )
        widgetView.setTextColor(
            R.id.textWidgetLocked,
            ColorUtility.applyTransparency(state.contentColor, state.contentAlpha)
        )
    }

    private fun getOnClickPendingIntent(
        context: Context,
        intentAction: String,
        widgetId: Int
    ): PendingIntent {
        val intent = Intent(context, this.javaClass)
        intent.action = intentAction
        intent.putExtra(KEY_WIDGET_ID, widgetId)
        return PendingIntent.getBroadcast(
            context,
            widgetId,
            intent,
            PendingIntent.FLAG_IMMUTABLE
        )
    }

    private fun log(context: Context, string: String) {
        CrystalNoteLogger.log(context, string)
    }


    /*--- Constants ---*/

    companion object {
        private const val TITLE_CLICK_INTENT = "com.xephorium.crystalnote.widget.click.TITLE"
        private const val TEXT_CLICK_INTENT = "com.xephorium.crystalnote.widget.click.TEXT"
        private const val EMPTY_CLICK_INTENT = "com.xephorium.crystalnote.widget.click.EMPTY"
        const val KEY_WIDGET_ID = "WIDGET_ID_KEY"

        fun refreshWidgets(context: Context, application: Application) {
            val intent = Intent(context, NotesWidgetProvider::class.java)
            val ids = AppWidgetManager.getInstance(application)
                    .getAppWidgetIds(ComponentName(application, NotesWidgetProvider::class.java))

            intent.action = AppWidgetManager.ACTION_APPWIDGET_UPDATE
            intent.putExtra(AppWidgetManager.EXTRA_APPWIDGET_IDS, ids)
            context.sendBroadcast(intent)
        }
    }
}